src/i_winmusic.c

```diff
@@ -1488,34 +1502,35 @@ static void I_WIN_SetMusicVolume(int volume)
 
     volume_factor = sqrtf((float)volume / 15);
 
-    update_volume = (song.file != NULL);
+    EnterCriticalSection(&CriticalSection);
+
+    win_midi_state = STATE_VOLUME;
+
+    LeaveCriticalSection(&CriticalSection);
 }
 
 static void I_WIN_StopSong(void *handle)
 {
     MMRESULT mmr;
 
-    if (!hPlayerThread)
+    if (!hMidiStream)
     {
         return;
     }
 
-    SetEvent(hExitEvent);
-    WaitForSingleObject(hPlayerThread, PLAYER_THREAD_WAIT_TIME);
-    CloseHandle(hPlayerThread);
-    hPlayerThread = NULL;
+    EnterCriticalSection(&CriticalSection);
+
     win_midi_state = STATE_STOPPED;
```

Comment:
When the song is stopped, `SendNotesSoundOff()` should be called since the `midiStreamStop()` function is insufficient by itself. Maybe add `SendNotesSoundOff()` to a new `STATE_STOPPING` (just an example name) case inside `FillBuffer()`. `midiStreamStop()` shouldn't be called until after that message is finished.

After these changes, `SendNotesSoundOff()` can be removed from `ResetDevice()`.

Reply:
I tried to do it in https://github.com/fabiangreffrath/woof/pull/1248/commits/4cf317980715b288be99c38e0bf77591bd550f06

Reply:
Please add `SendNotesSoundOff()` here since `I_WIN_StopSong(NULL)` is no longer called during shut down.

```diff
diff --git a/src/i_winmusic.c b/src/i_winmusic.c
index addb31b..963229a 100644
--- a/src/i_winmusic.c
+++ b/src/i_winmusic.c
@@ -1252,6 +1252,8 @@ static void FillBuffer(void)
             return;
 
         case STATE_SHUTDOWN:
+            // Send notes/sound off prior to reset to prevent volume spikes.
+            SendNotesSoundOff();
             ResetDevice();
             StreamOut();
             win_midi_state = STATE_EXIT;
```